unison-core/src/Unison/ABT.hs

```diff
@@ -244,12 +244,25 @@ into' a abt = case abt of
 
 -- | renames `old` to `new` in the given term, ignoring subtrees that bind `old`
 rename :: (Foldable f, Functor f, Var v) => v -> v -> Term f v a -> Term f v a
-rename old new t0@(Term _ ann t) = case t of
-  Var v -> if v == old then annotatedVar ann new else t0
-  Cycle body -> cycle' ann (rename old new body)
-  Abs v body -> if v == old then abs' ann v body
-                else abs' ann v (rename old new body)
-  Tm v -> tm' ann (fmap (rename old new) v)
+rename old new t0@(Term fvs ann t) =
+  if Set.notMember old fvs then t0
```

Comment:
This line is an optimization I didn't spot before - if the variable `old` isn't free in this subtree (either due to it being bound, or it just not appearing), the rename has no effect on this subtree, so skip it.

Reply:
Make sense to me. I'm surprised that wasn't in there, because I thought that was one of the main points of storing the free variables.

Reply:
Yeah it is done in the subst implementation but wasn’t being done here for some reason.